### PR TITLE
fix(re-work-logic): filename fetching logic to work with binaries

### LINKED ISSUE #2: Rewrite filename fetching logic to work with binaries
**Description:**
After releasing binaries for samosa through GoReleaser and GitHub Releases, the following error occurs on running the binary:

```
2022/03/26 09:38:54 failed to get coverage data: go/build: go list github.com/burntcarrot/samosa/internal/: exit status 2
go: cannot find GOROOT directory: /opt/hostedtoolcache/go/1.17.6/x64
```

**Pointers:**
The point of failure is inside [getFilename](https://github.com/burntcarrot/samosa/blob/main/internal/utils.go#L43). Investigate [getFilename](https://github.com/burntcarrot/samosa/blob/main/internal/utils.go#L43) and rewrite the function.

### PR DESCRIPTION
**Pull request by ehrktia**
🐛  fix module not found error
📝 updated README with todos
ℹ️  works for pkgs uploaded to version control like github for ex

Closes: #2.

### CODE REVIEW

**Comment by burntcarrot** [comment]
@ehrktia Thanks for the PR! Is this still in progress?

**Comment by ehrktia** [comment]
> @ehrktia Thanks for the PR! Is this still in progress?

Is finished can you please try a new module from your end before merge

**Comment by burntcarrot** [comment]
> Is finished can you please try a new module from your end before merge

Sorry, I haven't been getting time lately; I was busy with some other stuff. I'll test it on my end soon!

**Comment by ehrktia** [comment]
> > Is finished can you please try a new module from your end before merge
> 
> Sorry, I haven't been getting time lately; I was busy with some other stuff. I'll test it on my end soon!

🙇🏼‍♂️

File: internal/testdata/test_coverage.out [hunk 1]
```diff
@@ -1,3 +1,4 @@
 mode: set
+github.com/web-alytics/meditate/pkg/logging/logging.go:20.24,21.17 1 0
```
Comment (burntcarrot) [inline internal/testdata/test_coverage.out]: I think we would need to remove this. Since Github Actions would have access to the files of this repo (`burntcarrot/samosa`), it would be preferable to use them for testing.

File: internal/utils.go [hunk 2]
```diff
@@ -42,14 +45,49 @@ func filterByRegex(pattern string, fi []*funcInfo) ([]*funcInfo, error) {
...
+
+	} else {
+		pkg, err = build.ImportDir(dir, build.FindOnly)
+		if err != nil {
+			_, err = build.Import(dir, ".", build.FindOnly)
+			if err != nil {
+				if strings.Contains(err.Error(), "add") {
+					if err := importModule(strings.TrimSuffix(dir, "/")); err != nil {
+						return "", err
+					}
```
Comment (burntcarrot) [inline internal/utils.go]: Could you please explain these in simpler terms? How does the error checking happen here?

File: internal/utils.go [hunk 3]
```diff
@@ -42,14 +45,49 @@ func filterByRegex(pattern string, fi []*funcInfo) ([]*funcInfo, error) {
...
+			}
+			return "", err
+		}
+	}
 	return filepath.Join(pkg.Dir, file), nil
 }
 
+func importModule(s string) error {
+	cmd := exec.Command("go")
+	cmd.Args = append(cmd.Args, "get", fmt.Sprintf("%s@latest", s))
```
Comment (burntcarrot) [inline internal/utils.go]: I'm not sure on pinning dependencies to `@latest`. Is there any approach that we could take?

**Review by burntcarrot — COMMENTED** [review]
Thanks for the PR once again! I've added some comments; would love your thoughts.

File: internal/testdata/test_coverage.out (hunk 1 above)
Comment (ehrktia) [reply internal/testdata/test_coverage.out]: updated in https://github.com/burntcarrot/samosa/pull/4/commits/7ef6b0cf24cf128cd6d6aaaf9da75b06d632df4d

File: internal/utils.go (hunk 2 above)
Comment (ehrktia) [reply internal/utils.go]: added comments https://github.com/burntcarrot/samosa/pull/4/commits/31af7604145dd5a0bc3d1b54ddeab41b48a9394d commit as per request

File: internal/utils.go (hunk 3 above)
Comment (ehrktia) [reply internal/utils.go]: @burntcarrot not sure what is your thoughts on pinning version other than latest.

Comment (burntcarrot) [reply internal/utils.go]: Users can prefer to pin to a certain version for some dependencies; fetching the `latest` version can lead to the modification of `go.mod`; which is not desired in many situations.

The tool should not change the dependencies on its own - we should let the users decide what's best in their use case.

Comment (ehrktia) [reply internal/utils.go]: Agree will think about it and ping you

Comment (ehrktia) [reply internal/utils.go]: @burntcarrot sorry for the hibernation with PR , did get some time and worked it out, https://github.com/burntcarrot/samosa/pull/4/commits/3fec24c5b023c6275171e28d25f115d1fcd5b54b this is the latest version with all changes
refactoring.

File: cmd/samosa/main.go [hunk 4]
```diff
@@ -3,11 +3,11 @@ package main
 import (
 	"log"
 
-	"github.com/burntcarrot/samosa/command"
+	"github.com/ehrktia/samosa"
```
Comment (burntcarrot) [inline cmd/samosa/main.go]: Could you please change it back to `burntcarrot`?

**Review by burntcarrot — CHANGES_REQUESTED** [review]
@ehrktia Thanks for the work on the PR!

I'll test this once locally this week; and review the remaining files.

File: file_parse_test.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package samosa
+
+import (
+	"testing"
+)
+
+func Test_parse(t *testing.T) {
+	testOut := make(map[string]int)
+	err := readCoverReport("./coverage.out", testOut)
```
Comment (burntcarrot) [inline file_parse_test.go]: Oh, missed this! We would need to store a test coverage report in a directory called `testdata`. [Example](https://github.com/burntcarrot/samosa/blob/main/internal/testdata/test_coverage.out)

Comment (ehrktia) [reply file_parse_test.go]: updated in commit https://github.com/burntcarrot/samosa/pull/4/commits/3c915a4a20b36852fc8812ac8508012d78d766ec

File: cmd/samosa/main.go (hunk 4 above)
Comment (ehrktia) [reply cmd/samosa/main.go]: @burntcarrot am not sure how to achieve this coz my PR is refactored of the functions used to report coverage, in case I point to `github.com/burntcarrot/samosa` it will fetch old functions. 
Let us decide on this after testing.

Comment (burntcarrot) [reply cmd/samosa/main.go]: Sure. Keeping this open for future reference.

File: file_parse_test.go (hunk 5 above)
Comment (burntcarrot) [reply file_parse_test.go]: @ehrktia The tests fail because the testdata is not present in this PR, could you please add a coverage report in `testdata/coverage.out` with the content being like [this](https://github.com/burntcarrot/samosa/blob/main/internal/testdata/test_coverage.out)?

Comment (ehrktia) [reply file_parse_test.go]: @burntcarrot done https://github.com/burntcarrot/samosa/pull/4/commits/1581df96b4b282220ec55db8174bd1e846bc6d32

File: file_parse_test.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+		t.Fatal(err)
+	}
+}
+
+func Test_coverage(t *testing.T) {
+	inp := "25.31,27.2 4 1"
+	out := getCoverage(inp)
+	if out != 4 {
+		t.Fatalf("expected to get `4`got:%v", out)
+	}
```
Comment (burntcarrot) [inline file_parse_test.go]:
```suggestion
        want := 4
	got := getCoverage(inp)
	if got != want {
		t.Fatalf("got: %v, want:%v\n", got, want)
	}
```

File: file_walk.go [hunk 7]
```diff
@@ -130,73 +133,88 @@ func (v *Visitor) Visit(node ast.Node) ast.Visitor {
...
-	if err != nil {
-		return nil, 0, 0, err
-	}
-
-	fi, covered, total, err := getFunctionInfo(profiles)
+// walkDir gets list of go files in repo
+func walkDir() ([]string, error) {
+	var filesDir []string
+	// get mod dir
+	modeFile, err := getMod()
```
Comment (burntcarrot) [inline file_walk.go:187]: Could you please explain this? Is it related to `go.mod`?

File: file_walk_test.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package samosa
+
+import "testing"
+
+func Test_walk(t *testing.T) {
```
Comment (burntcarrot) [inline file_walk_test.go]: Sorry for the nitpick, the test names should be in the form of mixed case. Could you please rename the newly added tests to conform to this naming convention?

Reference: https://go.dev/doc/effective_go#mixed-caps

**Review by burntcarrot — CHANGES_REQUESTED** [review]
@ehrktia Just some minor comments added.

I'll test this once locally and once it's working as intended, you can proceed to perform cleanup. Thanks for your patience!

File: file_parse_test.go (hunk 5 above)
Comment (burntcarrot) [reply file_parse_test.go]: Thanks! Resolving this.

Comment (ehrktia) [reply file_parse_test.go]: added coverage in https://github.com/burntcarrot/samosa/pull/4/commits/1581df96b4b282220ec55db8174bd1e846bc6d32 commit

File: file_walk_test.go (hunk 8 above)
Comment (ehrktia) [reply file_walk_test.go]: updated in commit https://github.com/burntcarrot/samosa/pull/4/commits/b7d53ff413ea3f4c3776c3f6e49c28fd0f5e3c5a

File: file_walk.go (hunk 7 above)
Comment (ehrktia) [reply file_walk.go:187]: using `go env` and `GOMOD` the root directory for the current repo is being set and using the base path from the root of the directory using file walk func files in each dir is collected and used further, this will have the correct path as the root directory will be basepath of the repo.

**Review by burntcarrot — APPROVED** [review]
@ehrktia Sorry for the delay!

I tried testing the GoReleaser Github Actions workflow locally through [act](https://github.com/nektos/act), but I failed to retrieve the build artifacts. I finally ran GoReleaser locally, and it worked perfectly, so I believe we're good to go!

I'll be pushing some commits to this PR - you might need to wait for some time. There are a couple of tests that needs some more robust checking.

Also, there was a bug related to the `-p` flag: the `-p` flag is intended for displaying "package"-level filenames, but the PR reverted it back to fully-qualified filenames (absolute paths). I've fixed this and I will push a commit soon.

File: file_parse.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	"strings"
+)
+
+// read the file in lines and load in to map
+func readCoverReport(f string, covReport map[string]int) error {
+	file, err := os.Open(f)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
```
Comment (burntcarrot) [inline file_parse.go]: @ehrktia I'm confused on why do we need this file? I'm not seeing any references to these helper functions, except tests. Would you like me to proceed with removing this?

Comment (ehrktia) [reply file_parse.go]: @burntcarrot sorry we do not need this one , you can clean it up .

**Comment by burntcarrot** [comment]
@ehrktia Could you please create a test release on your fork? The test release will ensure that the GoReleaser action works as intended.

**Comment by ehrktia** [comment]
@burntcarrot you mean do you want me to create a release in my fork

**Comment by burntcarrot** [comment]
@ehrktia yes.

**Comment by ehrktia** [comment]
> @ehrktia yes.

@burntcarrot https://github.com/ehrktia/samosa/releases/tag/v0.0.0 - early release from fork as requested

**Comment by burntcarrot** [comment]
@ehrktia Seems like it didn't trigger the workflow. Could you please remove the current release (along with the tag), and create a new release with a major version? (eg. `v1.0.0`)

You can choose to create a normal release, instead of a pre-release.

**Comment by ehrktia** [comment]
> @ehrktia Seems like it didn't trigger the workflow. Could you please remove the current release (along with the tag), and create a new release with a major version? (eg. `v1.0.0`)
> 
> You can choose to create a normal release, instead of a pre-release.

@burntcarrot done https://github.com/ehrktia/samosa/releases/tag/v1.0.0

**Comment by burntcarrot** [comment]
> > @ehrktia Seems like it didn't trigger the workflow. Could you please remove the current release (along with the tag), and create a new release with a major version? (eg. `v1.0.0`)
> > You can choose to create a normal release, instead of a pre-release.
> 
> @burntcarrot done https://github.com/ehrktia/samosa/releases/tag/v1.0.0

Strange, the workflow should be triggered on forks too. I feel that we can handle this in a separate PR.

**Review by burntcarrot — APPROVED** [review]
@ehrktia Thank you for your patience! I'm merging this; thanks for your contribution! 🥳

I'll open up a couple of more issues related to planned features/fixes - please feel free to pick them up if you want to contribute to the project! 😄
